### PR TITLE
Add `vers` support in index_packages endpoint

### PR DESCRIPTION
**Pull request by keshav-space**
- resolves https://github.com/nexB/purldb/issues/155

### CODE REVIEW

**Review by pombredanne — COMMENTED** [review]
Thanks! this is looking good, but I would prefer we find a way to avoid code copy from VulnerableCode. How could we share this code?
And if we go ahead and merge this, what's the plan to remove this duplication?

**Comment by tdruez** [comment]
This PR is a blocker for https://github.com/nexB/scancode.io/pull/836#issuecomment-1692964810
@keshav-space @pombredanne what's the status here?

**Comment by keshav-space** [comment]
> This PR is a blocker for [nexB/scancode.io#836 (comment)](https://github.com/nexB/scancode.io/pull/836#issuecomment-1692964810) @keshav-space @pombredanne what's the status here?

@tdruez waiting for review from @pombredanne and @JonoYang

File: packagedb/package_managers.py
Comment (JonoYang) [inline packagedb/package_managers.py:598]: Add a new line here

File: packagedb/package_managers.py
Comment (JonoYang) [inline packagedb/package_managers.py:208]:
```suggestion
        Return the latest date from a list of mapping of PyPI ``downloads`` or  None.
```

File: packagedb/package_managers.py
Comment (JonoYang) [inline packagedb/package_managers.py:225]: I think its more Pythonic to sort the list of downloads by the upload time, then popping the latest download entry from the list.

```suggestion
        # Sort `downloads` from earliest "upload_time_iso_8601" to latest
        downloads = downloads.sort(key=lambda x: x["upload_time_iso_8601"])
        # Pop the latest download entry from `downloads`
        latest_download = downloads.pop()
        latest_date = latest_download.get("upload_time_iso_8601")
        if latest_date:
            latest_date = dateparser.parse(latest_date)
        return latest_date
```

File: packagedb/api.py
Comment (JonoYang) [inline packagedb/api.py:50]: Add a new line here

File: packagedb/package_managers.py
Comment (JonoYang) [inline packagedb/package_managers.py:1]: Bring the tests from https://github.com/nexB/vulnerablecode/blob/main/vulnerabilities/tests/test_package_managers.py over

**Review by JonoYang — CHANGES_REQUESTED** [review]
@keshav-space Thanks for the PR! I've left some comments for your consideration.

File: packagedb/package_managers.py
Comment (keshav-space) [reply packagedb/package_managers.py:225]: @JonoYang I was trying to keep `package_managers` code untouched and refactor it while migrating to fetchcode.

Comment (JonoYang) [reply packagedb/package_managers.py:225]: @keshav-space Ah, I see.

File: packagedb/api.py
Comment (pombredanne) [inline packagedb/api.py:404]: Make it more robust with:
```suggestion
        purls = request.data.get('packages') or []
```

File: packagedb/api.py
Comment (pombredanne) [inline packagedb/api.py:705]:
```suggestion
def resolve_versions(parsed_purl, vers):
```

File: packagedb/api.py
Comment (pombredanne) [inline packagedb/api.py:748]: Can we get this from univers instead?
 Either patch univers to expose a VERSION_CLASS_BY_PACKAGE_TYPE or provide a function there that returns a version class for a type. This should in all cases be computed and not hardcoded.

If you use the built-in mapping and use the version class attribute you will get what you need
https://github.com/nexB/univers/blob/205da7ecbf7b0f195662373ea710b2b84a877eb0/src/univers/version_range.py#L417 
```suggestion
VERSION_CLASS_BY_PACKAGE_TYPE = {pkg_type: range_class.version_class for pkg_type, range_class in RANGE_CLASS_BY_SCHEMES.items()}
```

```

**Review by pombredanne — CHANGES_REQUESTED** [review]
Thanks... some nits for your consideration.

**Review by JonoYang — APPROVED** [review]
@keshav-space LGTM! 

@pombredanne Is there anything else you see that needs to be updated?

**Comment by JonoYang** [comment]
@keshav-space 

Merging, thanks for the contribution!

**Pull request merged**
